app/Main.hs

```diff
@@ -1,7 +1,176 @@
-module Main (main) where
+{-# OPTIONS_GHC -Wno-incomplete-record-updates #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE PatternSynonyms #-}
+module Main (main, ClientInput (NewTxCBOR)) where
 
-import Control.Applicative (pure)
-import System.IO (IO)
+import Cardano.Api (AlonzoEra, AsType (AsAddressInEra, AsAlonzoEra, AsSigningKey, AsPaymentKey), NetworkId (Testnet), NetworkMagic (NetworkMagic), Lovelace (Lovelace), UTxO (UTxO), Tx, TxIn, deserialiseAddress, serialiseToCBOR, readFileTextEnvelope)
+import Control.Concurrent (forkIO, newChan, readChan, writeChan)
+import Control.Exception (SomeException, displayException, fromException, try)
+import Data.Aeson qualified as Aeson (FromJSON, ToJSON, encode, decodeStrict, toJSON)
+import Data.ByteString.Lazy qualified as ByteString (toStrict)
+import Data.Functor (void)
+import Data.Map qualified as Map (singleton)
+import Data.String (fromString)
+import Data.Text (Text)
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Data.Text.IO qualified as Text (putStrLn)
+import GHC.Generics (Generic)
+import Ledger (PubKeyHash)
+import Network.WebSockets (ConnectionException, receiveData, sendTextData, runClient)
+import System.Environment (getArgs)
+import System.IO.Error (isEOFError)
+
+import HydraRPS.OffChain (PlayParams (..))
+import HydraRPS.Tx (UserCredentials (..), mkUserCredentials, parseTxIn, txOutToAddress)
+
+import Prelude
 
 main :: IO ()
-main = pure ()
+main = do
+  nodeHost : nodePort : keyFile : _ <- getArgs
+  skey <- either (fail . show) pure =<< readFileTextEnvelope (AsSigningKey AsPaymentKey) keyFile
+  let networkId = Testnet (NetworkMagic 42)
+      userCreds = mkUserCredentials networkId skey
+  runClient nodeHost (read nodePort) "/" $ \ws -> do
+    let nextServerEvent = receiveData ws
+        submitCommand input = do
+          let json = Aeson.encode input
+          Text.putStrLn $ "client input: " <> decodeUtf8 (ByteString.toStrict json)
+          sendTextData ws json
+
+    events <- newChan
+    let enqueueApiEvent = writeChan events . ApiEvent
+        enqueuUserCommand = writeChan events . UserCommand
+        nextEvent = readChan events
+
+    void $ forkIO $ apiReader nextServerEvent enqueueApiEvent
```

Comment:
This would benefit from using [async](https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html)